client/src/util/centroid.js

```diff
@@ -6,100 +6,131 @@ import { unassignedCategoryLabel } from "../globals";
+/*
+Generates a mapping of categorical values to data needed to calculate centroids
+categoricalValue -> {
+ size: int,
+ holdsFinite: Boolean,
+ xCoordinates: Float32Array,
+ yCoordinates: F32 Array
+}
 */
-
-const calcMedianCentroid = (
+const getCoordinatesByCategoricalValues = (
   obsAnnotations,
   obsLayout,
   categoryName,
   layoutDimNames,
   categoricalSelection,
   schemaObsByName
 ) => {
+  const scratchPad = new Map();
+
   const categoryArray = obsAnnotations.col(categoryName).asArray();
 
   const layoutXArray = obsLayout.col(layoutDimNames[0]).asArray();
   const layoutYArray = obsLayout.col(layoutDimNames[1]).asArray();
-  const coordinates = new Map();
 
-  // Iterate over all the cells in the category
+  const { categoryValueIndices, categoryValueCounts } = categoricalSelection[
+    categoryName
+  ];
+
+  // Check to see if the current category is a user created annotation
+  const isUserAnno = schemaObsByName[categoryName].writable;
+
+  // Iterate over all cells
   for (let i = 0, len = categoryArray.length; i < len; i += 1) {
+    // Fetch the categorical value of the current cell
     const categoryValue = categoryArray[i];
 
     // Get the index of the categoryValue within the category
-    // If the category is truncated and this value is removed,
-    // it will not be assigned a category value and will not be
-    // labeled on the graph
-    const categoryValueIndex = categoricalSelection[
-      categoryName
-    ].categoryValueIndices.get(categoryValue);
-
-    // Check to see if the current category is a user created annotation
-    // if the user created this category, do not create a label for the `unassigned` value
-    const isUserAnno = schemaObsByName[categoryName].writable;
+    const categoryValueIndex = categoryValueIndices.get(categoryValue);
 
+    // If the category is truncated and this value is removed,
+    //  it will not be assigned a category value and will not be
+    //  labeled on the graph
+    // If the user created this category,
+    //  do not create a label for the `unassigned` value
     if (
       categoryValueIndex !== undefined &&
       !(isUserAnno && categoryValue === unassignedCategoryLabel)
     ) {
-      // Get the number of cells which are in the category value
-      const numInCategoryValue =
-        categoricalSelection[categoryName].categoryValueCounts[
-          categoryValueIndex
-        ];
-
-      // Create/fetch the valueArray,
-      // which is what the key points to in the `coordinates` hashmap
-      const valueArray = coordinates.get(categoryValue) || [
-        false, // hasFinite
-        0, // index
-        new Float32Array(numInCategoryValue), // x coordinates
-        new Float32Array(numInCategoryValue) // y coordinates
-      ];
-      const index = valueArray[1];
-      let hasFinite = valueArray[0];
-
-      hasFinite =
-        hasFinite ||
+      // Create/fetch the scratchpad value
+      let value = scratchPad.get(categoryValue);
```

Comment:
calling this variable `value` is confusing given all the other uses of the term.

perhaps:
* scratchpad becomes `coordsByCategoryLabel`
* value becomes `coords`
Or something like that.